File: src/main/kotlin/script/lua/library/Logger.kt
package org.lain.engine.script.lua.library

import org.lain.engine.Constants
import org.lain.engine.script.ScriptEngine
import org.lain.engine.script.lua.NIL
import org.lain.engine.script.lua.luaTable
import org.luaj.vm2.Varargs
import org.slf4j.spi.LoggingEventBuilder

fun debugScript(module: String, info: String) {
    if (Constants.DEBUG_ALL) {
        ScriptEngine.LOGGER.info("[$module] $info")
    }
}

private fun log(varargs: Varargs, builder: LoggingEventBuilder) {
    val str = varargs.arg(1).tojstring()
    val arguments = varargs.narg() - 1
    builder.setMessage(str)
    for (i in 1..arguments) {
        builder.addArgument(varargs.arg(i + 1).tojstring())
    }
    builder.log()
}

fun LoggerTable() = luaTable {
    function("info") { varargs ->
        log(varargs, ScriptEngine.LOGGER.atInfo())
        NIL
    }
    function("warn") { varargs ->
        log(varargs, ScriptEngine.LOGGER.atWarn())
        NIL
    }
    function("error") { varargs ->
        log(varargs, ScriptEngine.LOGGER.atError())
        NIL
    }
    function2("debug") { module, str ->
        debugScript(module.tojstring(), str.tojstring())
        NIL
    }
}

File: src/main/resources/defaults/scripts/core/meta/logger.lua
---@meta

--- Этот файл нужен для поддержки LuaLS-аннотаций.
--- Не импортируйте его, так как он может сломать поведение скриптов.

---@class Logger
local logger = {}

---@param str string
---@param ... string
function logger.info(str, ...) end

---@param str string
---@param ... string
function logger.warn(str, ...) end

---@param str string
---@param ... string
function logger.error(str, ...) end

---@param str string
---@param ... string
function logger.debug(str, ...) end